### PR TITLE
Always pass layout to inner blocks

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

This enables the layout config of a parent block to always be passed to its inner blocks by default. It's one of the remaining steps towards the stabilisation of layout support 🙂 

## Why?

Addresses [this comment](https://github.com/WordPress/gutenberg/pull/45326#discussion_r1006628243), as well as miscellaneous unexpected behaviour around the ability to set wide/full alignments on child blocks.

It also removes a bunch of logic that wasn't actually doing anything from some of the core block edit functions 😅 

### Main Features

* Makes layout available as context from the `BlockEdit` function;
* Retrieves layout from block edit context in `useInnerBlocksProps`;

### Cleanup

* Removes alignment logic from flow layout altogether;
* Fixes the issue reported in #43502 by adding a deprecation to the Query block;
* Removes all logic around passing layout to inner blocks from all the block edit functions.

## Testing Instructions

1. Test Buttons, Comments Pagination, Gallery, Group, Navigation, Post Content, Query, Query Pagination and Social Links blocks layout controls and make sure they still work as expected.
2. In the Group block, check that wide/full alignment appears for a child Image block by default;
3. In the same Group block, toggle "Inner blocks use content width" off. Check that wide/full controls no longer appear for the child Image block.
4. Check that the scenario described in #43502 is fixed.
5. In a Column block, toggle "Inner blocks use content width" on. A child Image block should now display wide/full alignment controls.
6. Check that children of Row block still show a "Width" control under dimensions and it still works as expected.

### CODE REVIEW

File: packages/block-library/src/query/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/query/deprecated.js:299]: Can we add a fixture to match this added deprecation? (e.g. `core__query__deprecated-4.html` etc)
It might be a little tricky if the deprecation depends on an earlier version of `useInnerBlocksProps.save()` to make sure that we test the deprecation is being run correctly 🤔

File: packages/block-library/src/query/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/query/deprecated.js]: By including `layout.contentSize` in the eligibility list, what happens if a user sets a content size after it's migrated to a constrained layout? Will this match happen another time?

**Review by andrewserong — COMMENTED** [review]
Yay, this is exciting, great to see this tidy up! I probably won't get a chance to give this a detailed look before I wrap up for the day, so just added a couple of drive-by comments about the query deprecation that stood out to me while skim-reading. Apologies if I've missed details / context there, as I was only looking quite quickly! 🙂

File: packages/block-library/src/query/deprecated.js
Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js:299]: Thanks, I always forget about the fixtures 😅

File: packages/block-library/src/query/deprecated.js
Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js]: That's a good question! I re-used the same logic from the Group block deprecation which seems to be working fine, but will double-check.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/inner-blocks/index.js]: Maybe we should use a stable object reference here. Otherwise, `useMemo` will return new values when the layout setting is "falsy".

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/inner-blocks/index.js]: Same here as `defaultLayoutBlockSupport` is a fallback for `usedLayout`.

File: packages/block-library/src/query/deprecated.js
Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js]: Yeah, turns out it does run so I added an extra check to only run it if content size is set _without_ constrained layout.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/inner-blocks/index.js]: Done!

**Comment by tellthemachines** [comment]
Thanks for reviewing, folks! I've addressed the feedback so far and added a deprecation fixture so I think this is ready for another round.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (Mamaduka) [inline packages/block-editor/src/components/inner-blocks/index.js]: We should move this outside the component to avoid re-creating variables on every re-render. You can find similar instances of `EMPTY_OBJECT` in the codebase.

Comment (tellthemachines) [reply packages/block-editor/src/components/inner-blocks/index.js]: Oh, got it, thanks!

Comment (tellthemachines) [reply packages/block-editor/src/components/inner-blocks/index.js]: Done!

File: packages/block-editor/src/components/block-edit/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/block-edit/index.js:31]: This might be super edge-casey, but I was wondering if we should check whether or not the block has opted-in to the layout support before passing down its `layout` attribute in the context? Because this will apply to all blocks, if a 3rd party block happens to use an attribute also named `layout`, then it will be passed down here, too, without a check for the layout block support.

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (andrewserong) [inline packages/block-editor/src/components/inner-blocks/index.js:100]: At the moment, expanding `...defaultLayout` gives the passed down layout object access to the theme's content/wide size, so I think this sounds pretty good to me. But, looking at this raises a couple of potential questions (not blockers, just questions):

* Will the child always want content/wide size from the theme? Or to put it differently, will children ever need to know where the value comes from (i.e. parent vs theme)
* Will the child always want all the attributes from the default layout object, or is it just the content/wide size? Settings objects don't always have the same shape as objects in block attributes... but maybe they should match? If we're concerned about this, one option might be to just pick content/wide size here rather than passing everything to the `memoedLayout`.
* The default layout object also includes layout definitions, which could be really useful to child, or might not be needed potentially. Does it matter that this is merged into `memoedLayout`? I don't have a strong opinion either way here, mostly just curious if it's something we'll need to keep in mind.

File: packages/block-editor/src/layouts/flow.js
Comment (andrewserong) [inline packages/block-editor/src/layouts/flow.js:70]: Just double-checking — is the idea that this can be removed now because of the layout object being passed down? From testing, the issue of the wide width post template showing up using content size appears to have returned in this PR. I think the flow layout might still need the logic here (or elsewhere), so that the `alignwide` and `alignfull` classnames can be injected.

With TwentyTwentyTwo theme applied, it looks like the post template is being rendered as wide width on the site frontend, but content width in the site editor:

| Editor (note 650px) | Site frontend (note 1000px) |
| --- | --- |
| image | image |

It looks like on the site frontend the post template block's output has both `is-layout-flow` and `alignwide` classes as expected, however in the site editor the `alignwide` class is not present.

File: test/integration/fixtures/blocks/core__query.serialized.html
Comment (andrewserong) [inline test/integration/fixtures/blocks/core__query.serialized.html:1]: I couldn't quite work out why this change to serialized markup for the block with empty attributes is required. Is it because of the deprecation updating the attributes? In practice I don't think it really matters that this serializes the default values of the empty state of the block, but was just curious as to why that's happening now 🤔

**Review by andrewserong — COMMENTED** [review]
Thanks for all the follow-up here @tellthemachines! I haven't finished working through testing all the blocks, but just thought I'd leave the review now as I think the change to the full/wide alignments for flow layout appears to reintroduce the issue that was resolved by https://github.com/WordPress/gutenberg/pull/43502. Let me know if you need any other info there!

Otherwise, I like the logic change here, and it feels like it's honing in on a good structure for passing down layout in an intuitive way that results in blocks needing to do less work — it seems like this is nearly there!

**Comment by tellthemachines** [comment]
> I think the change to the full/wide alignments for flow layout appears to reintroduce the issue that was resolved by https://github.com/WordPress/gutenberg/pull/43502. Let me know if you need any other info there

Funny, it's working for me. What happens when you paste the test markup into the code editor? Doesn't the Query block layout attribute get auto-updated with `"type":"constrained"`?

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/inner-blocks/index.js:100]: Good questions!

> Will the child always want content/wide size from the theme? Or to put it differently, will children ever need to know where the value comes from (i.e. parent vs theme)

If the parent also has a content or wide size set, it will override the theme one because we're adding `usedLayout` under `defaultLayout`.

> Will the child always want all the attributes from the default layout object, or is it just the content/wide size?

> The default layout object also includes layout definitions, which could be really useful to child, or might not be needed potentially.

We don't know what we might yet need 😄 but given we're passing layout definitions around pretty much everywhere in layout, this just sticks to the same behaviour. The general idea is that `BlockListItems` should be able to access anything from layout.

File: packages/block-library/src/query/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/query/deprecated.js:367]: Just noticed that if `layout.inherit` is `true`, then this still gets included in the object here. Should we add a `inherit: undefined` line here after `...layout`?

File: packages/block-editor/src/layouts/flow.js
Comment (tellthemachines) [reply packages/block-editor/src/layouts/flow.js:70]: That code needs to be removed because it's causing align wide and full controls to _always_ show, even if the parent block doesn't have a content width. The problem is that the layout we're extracting content and wide size from there is the full theme layout, not the block one, so it always includes those properties if they're set by the theme. 

The specific problem those lines were fixing should now be fixed by the Query block deprecation. I only checked with the test markup from your PR, but will have a look at TT2 now.

**Review by andrewserong — COMMENTED** [review]
Oh, interesting, yes if I cut the markup from TT2's home template so that there's a change to the site's markup and then paste it back in within the code editor view then the deprecation runs. However, for the template that ships with the theme, when opening the site editor, it doesn't look like the query block is updated immediately 🤔

This is the test markup from TT2:

<details>

<summary>Test markup from TT2's home template</summary>

```
<!-- wp:query {"queryId":0,"query":{"perPage":10,"pages":0,"offset":0,"postType":"post","order":"desc","orderBy":"date","author":"","search":"","exclude":[],"sticky":"","inherit":true},"tagName":"main","displayLayout":{"type":"list"},"layout":{"inherit":true}} -->
<main class="wp-block-query"><!-- wp:post-template {"align":"wide"} -->
<!-- wp:group {"layout":{"inherit":true,"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:post-title {"isLink":true,"align":"wide","fontSize":"var(\u002d\u002dwp\u002d\u002dcustom\u002d\u002dtypography\u002d\u002dfont-size\u002d\u002dhuge, clamp(2.25rem, 4vw, 2.75rem))"} /-->

<!-- wp:post-featured-image {"isLink":true,"align":"wide","style":{"spacing":{"margin":{"top":"calc(1.75 * var(\u002d\u002dwp\u002d\u002dstyle\u002d\u002dblock-gap))"}}}} /-->

<!-- wp:columns {"align":"wide"} -->
<div class="wp-block-columns alignwide"><!-- wp:column {"width":"650px"} -->
<div class="wp-block-column" style="flex-basis:650px"><!-- wp:post-excerpt /-->

<!-- wp:post-date {"format":"F j, Y","isLink":true,"style":{"typography":{"fontStyle":"italic","fontWeight":"400"}},"fontSize":"small"} /--></div>
<!-- /wp:column -->

<!-- wp:column {"width":""} -->
<div class="wp-block-column"></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:spacer {"height":"112px"} -->
<div style="height:112px" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer --></div>
<!-- /wp:group -->
<!-- /wp:post-template -->

<!-- wp:query-pagination {"paginationArrow":"arrow","align":"wide","layout":{"type":"flex","justifyContent":"space-between"}} -->
<!-- wp:query-pagination-previous {"fontSize":"small"} /-->

<!-- wp:query-pagination-numbers /-->

<!-- wp:query-pagination-next {"fontSize":"small"} /-->
<!-- /wp:query-pagination --></main>
<!-- /wp:query -->
```

</details>

So what I'm wondering:

* Is there some reason why the deprecation isn't initially running when the site editor opens the template, either in the deprecation, or in the site editor?
* If it's tricky to figure out, would it be worth retaining the wide / full alignment logic in the `layouts/flow.js` file for backwards compatibility, or does it cause other issues keeping it in place? 🤔

File: packages/block-editor/src/components/inner-blocks/index.js
Comment (andrewserong) [reply packages/block-editor/src/components/inner-blocks/index.js:100]: > If the parent also has a content or wide size set, it will override the theme one because we're adding usedLayout under defaultLayout.

Yeah, I think what I was curious about was whether or not the child will ever need to know where it came from (if the value was explicitly set on the parent, or if it's inherited from the default layout object). It's probably not something to worry about though, because we intentionally removed that complexity by getting rid of the `inherit` prop. I imagine it'll only be something to look into if we ever have a use case for it, which seems unlikely 🙂

> We don't know what we might yet need 😄 but given we're passing layout definitions around pretty much everywhere in layout, this just sticks to the same behaviour. 

Sounds good. And if we stick to passing everything down, then we've got consistent behaviour for if/when additional settings get added in `theme.json`, so it shouldn't be a surprise 👍

I'll resolve this thread.

File: packages/block-library/src/query/deprecated.js
Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js:367]: I don't think there's any harm in keeping it. We've been keeping it for the Group block deprecation, and on the whole I think the less we change the less scope there is for things to go wrong 😅

**Comment by tellthemachines** [comment]
> Is there some reason why the deprecation isn't initially running when the site editor opens the template, either in the deprecation, or in the site editor?

Yeah, that's a very good question 😅 

I'm noticing further that the deprecation only _doesn't_ run if the template has no customizations, so there must be some check that skips brand new templates. That means we can solve the problem by updating the theme markup in Core, because any instance of the template already in use is likely to have some customization to it so should work. 

We should get to the bottom of the deprecation issue, but I'm hesitant to block this change for its sake, so suggest we do it as a follow-up task.

> If it's tricky to figure out, would it be worth retaining the wide / full alignment logic in the layouts/flow.js file for backwards compatibility, or does it cause other issues keeping it in place?

It does cause other issues - currently we have children of Group blocks with flow layout showing wide/full controls, which feels broken because those controls don't do anything if the parent is a flow block.

File: packages/block-library/src/query/deprecated.js
Comment (andrewserong) [reply packages/block-library/src/query/deprecated.js:367]: The main reason I thought of removing it is that a truthy check for `layout.inherit` is included in the `isEligible` callbacks, so I was wondering if it'd be possible for a migrated block to be treated as though it hasn't been migrated if the inherit value sticks around 🤔

Totally agree with trying to change fewer things, though!

**Comment by andrewserong** [comment]
> We should get to the bottom of the deprecation issue, but I'm hesitant to block this change for its sake, so suggest we do it as a follow-up task.

Sounds good — if I go in to the template and make a change, e.g. adding a paragraph and then save and reload the site editor, then the migration happens. So it's fairly easy for a template to get to the correct state, it's just unfortunate that that isn't happening earlier 🤔

> It does cause other issues - currently we have children of Group blocks with flow layout showing wide/full controls, which feels broken because those controls don't do anything if the parent is a flow block.

I always find these ones tricky where we need to choose between two things that might feel broken, and determine which is the less broken-feeling one. But in this case, and with the timeline on feature freeze for 6.2, I think I agree with the logic here 👍

I'll continue working through the other tests cases!

File: packages/block-editor/src/components/block-edit/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/block-edit/index.js:31]: Good idea, let's do that!

File: test/integration/fixtures/blocks/core__query.serialized.html
Comment (tellthemachines) [reply test/integration/fixtures/blocks/core__query.serialized.html:1]: Oooh that's weird, not sure why it's happening 🤔

Comment (tellthemachines) [reply test/integration/fixtures/blocks/core__query.serialized.html:1]: I simplified the fixture a bit and removed some values I'd incorrectly copied from the previous deprecation, but no changes to this file. I guess it should be fine? 🤷

Comment (andrewserong) [reply test/integration/fixtures/blocks/core__query.serialized.html:1]: Good idea simplifying the fixtures. For this one, I think it's just that what gets serialized for the block's default state has changed — I don't think it's really an issue though since it's just the block attribute's metadata for when it's in the placeholder state. I'll just CC @ntsekouras for visibility, but I don't think it'd be a blocker.

**Review by andrewserong — APPROVED** [review]
Alrighty, this is testing well for me, the only issues I ran into are the ones we've already discussed, and it seems to me that the trade-offs favour merging in the current state, as it's a big improvement to consolidate the logic surrounding passing down layout, and it's a great idea getting this in for 6.2.

I tested the following:

* The following blocks behave correctly in the editor and render correctly on the site frontend as far as I can tell:
  * ✅ Buttons
  * ✅ Comments Pagination
  * ✅ Group
  * ✅ Navigation
  * ✅ Post Content
  * ✅ Query
  * ✅ Query Pagination
  * ✅ Social Links
* ✅ Tested group block as wrapper for image block, and that wide/full alignement controls only appear for constrained layout type
* ❓ Checked that the scenario described in [Layout: Re-instate alignwide and alignfull in flow layout get alignments #43502](https://github.com/WordPress/gutenberg/pull/43502) is fixed — it is fixed once the deprecation is run, but in TT2 where the template does not have the latest markup, the deprecation only appears to run when the template is altered. I agree that it's not a blocker for this PR to land, but it'd be a good follow up bug to look into.
* ✅ Column block child image only has wide/full controls when constrained layout is set
* ✅ Row children width controls are working correctly

Nice work, LGTM! ✨

File: packages/block-library/src/query/deprecated.js
Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js:367]: We can also change the check to not run at all if `type: "contstrained"` is present.

**Comment by tellthemachines** [comment]
Thanks for the reviews and extensive testing @andrewserong 🎉 I'm going to merge this as is to unblock #47584 and hopefully also #45326, and start looking into the follow-ups.

File: packages/block-library/src/query/deprecated.js
Comment (ntsekouras) [reply packages/block-library/src/query/deprecated.js:367]: Why other blocks don't have a deprecation except for Query? It's weird to have an empty Query Loop with all the attributes set.. Also the `migrate` function should run only to this deprecation? Usually it's needed to also run for [other migrations](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/)..

**Comment by corentin-gautier** [comment]
Hi @tellthemachines I no longer have wide/full alignment options on first level blocks since this has been merged, they do appear on child blocks though

File: packages/block-library/src/query/deprecated.js
Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js:367]: The Query deprecation doesn't address this change, it should have been added as part of #42763. I added it here in an attempt to fix the issue reported in #43502, as the solution from that PR was causing other breakage.

> Also the migrate function should run only to this deprecation? Usually it's needed to also run for other migrations...

I'm not sure I understand. Is there anything I should change here?

**Comment by tellthemachines** [comment]
Hi @corentin-gautier, could you provide some more details to help debug?

* Is the issue occurring in the Post editor, the Site editor or both?
* Which theme are you using? 
* Can you reproduce the issue with a default theme (Twenty Twenty Three or Twenty Twenty Two)?

Thanks!

**Comment by corentin-gautier** [comment]
Hi @tellthemachines 

I'm using a custom theme which doesn't use the site editor so I only experienced this issue in the post editor, the issue doesn't happen in twenty twenty three

EDIT: It does happen in twenty twenty three, but after a few seconds it goes back to normal

https://user-images.githubusercontent.com/4660731/216286991-01f11674-5b50-4322-817d-e680da3cbe3c.mov

**Comment by corentin-gautier** [comment]
@tellthemachines Hum, I just re-tested this morning and I don't have the issue anymore

File: packages/block-library/src/query/deprecated.js
Comment (ntsekouras) [reply packages/block-library/src/query/deprecated.js:367]: I haven't tested, but this is from the [docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/):
>It is important to note that if a deprecation’s save method does not produce a valid block then it is skipped completely, including its migrate method, even if isEligible would return true for the given attributes. This means that if you have several deprecations for a block and want to perform a new migration, like moving content to InnerBlocks, you may need to update the migrate methods in multiple deprecations in order for the required changes to be applied to all previous versions of the block.

You could test by copy/paste the other deprecated versions and see if there are errors.

Comment (tellthemachines) [reply packages/block-library/src/query/deprecated.js:367]: Oh, got it, thanks! I'll have a look.

**Comment by gyurmey2** [comment]
Hi @tellthemachines, @corentin-gautier, I am facing a similar problem. For some reason, it's only found in some themes, e.g. Upsidedown & Disco By Automatic, but there are more of these themes. [You can test at this link.](https://app.instawp.io/wordpress-auto-login?site=%242y%2410%24gfPkC9IKvAjGCZBgpoW7JeEfTC8JJ2qt.z1kpcBY4iNrineDPAvFi)

https://user-images.githubusercontent.com/113413673/217945392-4e871ce8-73e5-45df-b5ab-3585d29a44cb.mp4

**Comment by tellthemachines** [comment]
Hi @gyurmey2 , thanks for the report! I can confirm that on that test site, with Upsidedown theme, the wide/full controls don't appear for root blocks in the post editor. This happens because the Post Content block in the Single template doesn't have content width enabled:

Screenshot 2023-02-10 at 9 36 36 am

You'll notice that if you disable the Gutenberg plugin, root blocks in the post editor now show the wide/full controls:

Screenshot 2023-02-10 at 9 34 36 am

However, on the front end, those blocks set to wide/full are the same width as regular blocks:

Screenshot 2023-02-10 at 9 33 10 am

So the change that this PR introduced brings consistency between editor and front end, in that if the Post Content doesn't set a content width for its children, they will always take up the full width of their container (which in this case is the Post Content block), so wide/full controls aren't available because they won't have any effect.

@corentin-gautier I suspect something similar might be happening with your custom theme. The issue that your video captures in Twenty Twenty Three is slightly different, as that theme does have content width in its Post Content block. That layout change we can see in the video is due to a performance issue with loading post content styles in the post editor, which #45299 should fix.

**Comment by Mamaduka** [comment]
@tellthemachines, I think this report is also related to changes here - #47857.

Edit: Confirmed by using `git bisect` that the Hybrid theme regression commit is fff6d811cef2944935dca3166a9936da5a335428.

**Comment by gyurmey2** [comment]
@tellthemachines, thank you very much for the detailed explanation (now I understand why this is happening).

Another problem I've been struggling with for a long time is the [differences between the editor and the frontend](https://rikes-back-pamu.instawp.xyz/test/). Could you help please? [Same test link](https://app.instawp.io/wordpress-auto-login?site=%242y%2410%24gfPkC9IKvAjGCZBgpoW7JeEfTC8JJ2qt.z1kpcBY4iNrineDPAvFi).

**Comment by corentin-gautier** [comment]
@tellthemachines Thing is : I'm not using the template editor in my theme (I use classic php > twig files) which means I don't have a post content block because I don't have a Single Template. Which means this PR breaks the alignment on root blocks for classic themes ... 

It's exactly what's described in #47857 

EDIT: Oh I see there's a PR for this already : #47961

**Comment by tellthemachines** [comment]
@corentin-gautier I see, thanks for flagging this! It should be fixed with #47961.

**Comment by tellthemachines** [comment]
> Another problem I've been struggling with for a long time is the [differences between the editor and the frontend](https://rikes-back-pamu.instawp.xyz/test/). Could you help please? [Same test link](https://app.instawp.io/wordpress-auto-login?site=%242y%2410%24gfPkC9IKvAjGCZBgpoW7JeEfTC8JJ2qt.z1kpcBY4iNrineDPAvFi).

@gyurmey2 that is an actual bug! Looks like it has already been reported in #46724. I'll leave a comment over there to keep everything in the same place.

**Pull request merged**
